module/platform/Webots/src/Webots.cpp

```diff
@@ -293,168 +299,197 @@ namespace module::platform {
-            }
-            // If we don't have any data then this should correspond to a message with all fields set to defaults
-            else {
-                translate_and_emit_sensor(SensorMeasurements(SensorMeasurements::protobuf_type()));
-            }
-
-            // Service the watchdog
-            emit<Scope::WATCHDOG>(ServiceWatchdog<Webots>());
-        });
+            // Receiving
+            read_io = on<IO>(fd, IO::READ | IO::CLOSE | IO::ERROR).then("Read Stream", [this](const IO::Event& event) {
+                if ((event.events & IO::READ) != 0) {
+                    // Service the watchdog
+                    emit<Scope::WATCHDOG>(ServiceWatchdog<Webots>());
+
+                    // Work out how many bytes are available to read in the buffer and ensure we have enough space to
+                    // read them in our data buffer
+                    unsigned long available = 0;
+                    if (::ioctl(fd, FIONREAD, &available) < 0) {
+                        log<NUClear::ERROR>(fmt::format("Error querying for available data, {}", strerror(errno)));
+                        return;
+                    }
+                    buffer.reserve(buffer.size() + available);
+
+                    // Read data into our buffer and resize it to the new data we read
+                    auto bytes_read = ::read(fd, buffer.data() + buffer.size(), buffer.capacity() - buffer.size());
+                    buffer.resize(buffer.size() + bytes_read);
+
+                    // Function to read the payload length from the buffer
+                    auto read_length = [](const std::vector<uint8_t>& buffer) {
+                        return buffer.size() >= sizeof(uint32_t)
+                                   ? ntohl(*reinterpret_cast<const uint32_t*>(buffer.data()))
+                                   : 0;
+                    };
+
+                    // So long as we have enough bytes to process an entire packet, process the packets
+                    for (uint32_t length = read_length(buffer); buffer.size() >= length + sizeof(length);
+                         length          = read_length(buffer)) {
+                        // Decode the protocol buffer and emit it as a message
+                        char* payload = reinterpret_cast<char*>(buffer.data()) + sizeof(length);
+
+                        // TODO do something here with this payload and length to extract the protocol buffer and emit
+                        // it
+                        log<NUClear::TRACE>("Received sensor measurements");
+                        translate_and_emit_sensor(
+                            NUClear::util::serialise::Serialise<SensorMeasurements>::deserialise(payload, length));
+
+                        // Delete the packet we just read ready to read the next one
+                        buffer.erase(buffer.begin(), std::next(buffer.begin(), sizeof(length) + length));
```

Comment:
How does this affect size and capacity?

Reply:
@Bidski @TrentHouliston I'm hoping one of you can help here

Reply:
Erase will reduce the size, but I believe it will not reduce the capacity